### PR TITLE
Loading multichannel detector data using Databroker

### PR DESCRIPTION
**Pull request by dmgav**
Fixed multiple issues related to loading of experimental data using Databroker. The main issues:

- Function ```make_hdf``` was not considering 3-channel data from Xpress3 detector as 1-channel data. When computing the sum of channels, only data from channel 1 was used. The fixed code correctly determines the number of channels in the input stream (using table dimensions) and uses complete data for calculations. Changes are only related to fly scan mode on the SRX beamline;

- Function ```make_hdf``` was crashing in attempt to load step-scan data from Databroker. The fixed code allows to load step-scan data from one or two Xpress 3 detectors. Code changes are only relevant for step scan data from SRX beamline.

Testing was performed on SRX WS4.

### CODE REVIEW

File: pyxrf/model/load_data_from_db.py [hunk 1]
```diff
@@ -43,6 +43,46 @@
...
+            new_data[1::2, :, :] = new_data[1::2, ::-1, :]
+        else:
+            i = 0
+            for nx, ny in subscan_dims:
+                start = i + 1
+                end = i + ny
+                new_data[start:end:2, :, :] = new_data[start:end:2, ::-1, :]
+                i += ny
+    return new_data
+
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]:
```suggestion
```

File: pyxrf/model/load_data_from_db.py [hunk 2]
```diff
@@ -282,8 +322,18 @@ def map_data2D_srx(runid, fpath,
...
+        print(f"****************************************")
+        print(f"        Loading SRX step scan           ")
+        print(f"****************************************")
+
         fly_type = None
 
+        if num_end_lines_excluded is None:
+            datashape = [start_doc['shape'][1], start_doc['shape'][0]]   # vertical first then horizontal, assuming fast scan on x
+        else:
+            datashape = [start_doc['shape'][1]-num_end_lines_excluded, start_doc['shape'][0]]
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]:
```suggestion
            datashape = [start_doc['shape'][1] - num_end_lines_excluded, start_doc['shape'][0]]
```

File: pyxrf/model/load_data_from_db.py [hunk 3]
```diff
@@ -306,24 +356,27 @@ def map_data2D_srx(runid, fpath,
...
                             pos_list=hdr.start.motors,
                             scaler_list=config_data['scaler_list'],
                             fly_type=fly_type,
                             base_val=config_data['base_value'])  #base value shift for ic
             if 'xs2' in hdr.start.detectors:
                 print('Saving data to hdf file for second xspress3 detector.')
-                tmp = fpath.split('.')
-                fpath1 = '.'.join([tmp[0]+'_1', tmp[1]])
+                root, ext = os.path.splitext(fpath)
+                fpath1 = ''.join([root+'_1', ext])
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]: Can we use f-strings (or `.format()`) here instead?
```suggestion
                fpath1 = f"{root + '_1'}{ext}"
```

File: pyxrf/model/load_data_from_db.py [hunk 4]
```diff
@@ -335,16 +388,25 @@ def map_data2D_srx(runid, fpath,
...
+        print(f"****************************************")
+        
         if save_scalar is True:
             scaler_list = ['i0', 'time', 'i0_time', 'time_diff']
             xpos_name = 'enc1'
             ypos_name = 'hf_stage_y' # 'hf_stage_x' if fast axis is vertical
+
+        num_det = 0  # Some default value (should never be used)
+
+
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]:
```suggestion
```

File: pyxrf/model/load_data_from_db.py [hunk 5]
```diff
@@ -335,16 +388,25 @@ def map_data2D_srx(runid, fpath,
...
-        if 'E_tomo' in start_doc['scaninfo']['type']:
-            num_det = 1
-            ypos_name = 'e_tomo_y'
-        else:
-            num_det = 3
+        #if 'E_tomo' in start_doc['scaninfo']['type']:
+        #    num_det = 1
+        #    ypos_name = 'e_tomo_y'
+        #else:
+        #    num_det = 3
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]: Why was this block commented out?

File: pyxrf/model/load_data_from_db.py [hunk 6]
```diff
@@ -378,13 +440,21 @@ def map_data2D_srx(runid, fpath,
...
+            if m == 0:
+                # Determine the number of channels from the size of the table with fluorescence data
+                num_det = v.data['fluor'].shape[1]
+
+                # Now allocate space for fluorescence data
+                if create_each_det is False:
+                    new_data['det_sum'] = np.zeros(new_shape)
+                else:
+                    for i in range(num_det):
+                        new_data['det'+str(i+1)] = np.zeros(new_shape)
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]:
```suggestion
                        new_data[f'det{i + 1}'] = np.zeros(new_shape)
```

File: pyxrf/model/load_data_from_db.py [hunk 7]
```diff
@@ -397,6 +467,10 @@ def map_data2D_srx(runid, fpath,
                             interp_list = (v.data[n][-1] - v.data[n][-3]) / 2 * np.arange(1, len_diff + 1) + v.data[n][-1]
                             data[n][m, min_len:datashape[1]] = interp_list
                 fluor_len = v.data['fluor'].shape[0]
+                if m > 0 and not (m % 10):
+                    print(f"Processed #{m} lines ...")
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]: Would it make sense to also add a total number of lines in the print?

File: pyxrf/model/load_data_from_db.py [hunk 8]
```diff
@@ -406,6 +480,14 @@ def map_data2D_srx(runid, fpath,
                         # in case the data length in each line is different
                         new_data['det'+str(i+1)][m, :fluor_len, :] = v.data['fluor'][:,i,:]
 
+        # Modify file name (path) to include data on how many channels are included in the file and how many
+        #    channels are used for sum calculation
+        root, ext = os.path.splitext(fpath)
+        s = f"_sum({num_det}ch)"
+        if create_each_det:
+            s += f"+{num_det}ch"
+        fpath = ''.join([root+s, ext])
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]:
```suggestion
        fpath = f'{root + s}{ext}'
```

File: pyxrf/model/load_data_from_db.py [hunk 9]
```diff
@@ -432,7 +514,9 @@ def map_data2D_srx(runid, fpath,
             data1 = db.get_table(hdr, fill=True, stream_name='primary')
             if num_end_lines_excluded is not None:
                 data1 = data1[:datashape[0]]
-            if ypos_name not in data1.keys() and 'E_tomo' not in start_doc['scaninfo']['type']:
+            # if ypos_name not in data1.keys() and 'E_tomo' not in start_doc['scaninfo']['type']:
+            #print(f"data1 keys: {data1.keys()}")
```
Comment (mrakitin) [inline pyxrf/model/load_data_from_db.py]: Can those debug prints be removed?
```suggestion
```

File: pyxrf/model/load_data_from_db.py (hunk 5 above)
Comment (dmgav) [reply pyxrf/model/load_data_from_db.py]: This condition is no longer valid. I am keeping the commented code for reference for now.

File: pyxrf/model/load_data_from_db.py (hunk 8 above)
Comment (dmgav) [reply pyxrf/model/load_data_from_db.py]: ```fstring``` will work as well

File: pyxrf/model/load_data_from_db.py (hunk 9 above)
Comment (dmgav) [reply pyxrf/model/load_data_from_db.py]: This is work in progress, so I would like to keep the debug prints for now. This code will have to be rewritten in the nearest future anyway and then this functionality will be moved from PyXRF to respective suitcase(s).

File: pyxrf/model/load_data_from_db.py (hunk 1 above)
Comment (dmgav) [reply pyxrf/model/load_data_from_db.py]: I think the permanent fix to formatting issues in PyXRF is to include flake8 check in travis built, but it requires reformatting of all PyXRF code.

File: pyxrf/model/load_data_from_db.py (hunk 7 above)
Comment (mrakitin) [reply pyxrf/model/load_data_from_db.py]: We discussed it in person. It's looping over a generator of events, so we don't directly know the length of it. Can be implemented in a separate PR later.

File: pyxrf/model/load_data_from_db.py (hunk 1 above)
Comment (mrakitin) [reply pyxrf/model/load_data_from_db.py]: Agree. It's worth to do it via a separate PR as it's a major update.

**Pull request merged**